En primer lugar, para comenzar el proyecto instalamos todas las librerias y paquetes necesarios, en donde las más utilizadas serán numpy que nos permite realziar operaciones matemáticas de arreglos y matrices, pandas que permite gestionar dataframes y sklearn que permite gestionar preprocesamiento y algoritmos de machine learning.

In [3]:
import numpy as np

In [4]:
import matplotlib

In [5]:
import pandas as pd
import scipy

In [6]:
import sklearn

In [7]:
import os


Una vez instaladas las librerías se procede a descargar la información csv, descargada en mi computador, únicamente se descargo la información de los tweets y no la correspondiente a la latitud y longitud de donde fueron realizados los mismos, ya que como se mencionará más adelante, esta no se considera necesaria. Una vez descargados los datos se verifica que hayan quedado guardados correctamente.

In [8]:
pwd

'C:\\Users\\anmam'

In [9]:
datos=pd.read_csv(r"C:\Users\anmam\OneDrive\Documentos\Ejercicio Twitter\auspol2019.csv")

In [10]:
datos.head()

created_at                   id  \
0  2019-05-20 09:13:44  1130401208756187136   
1  2019-05-20 09:13:43  1130401205367140357   
2  2019-05-20 09:13:33  1130401162782371841   
3  2019-05-20 09:13:29  1130401143551434753   
4  2019-05-20 09:13:23  1130401118666809345   

                                           full_text  retweet_count  \
0  After the climate election: shellshocked green...            0.0   
1  @narendramodi @smritiirani Coverage of indian ...            0.0   
2  @workmanalice Do you know if Facebook is relea...            0.0   
3  @vanbadham We all understand we have a compuls...            0.0   
4  Shares were mixed in Asia, with India and Aust...            0.0   

   favorite_count       user_id          user_name user_screen_name  \
0             0.0  9.248486e+07       PIPELINEPETE         jocksjig   
1             0.0  7.756474e+08    Narinder Parmar      nparmar1957   
2             0.0  5.687300e+04        Peter Wells       peterwells   
3             0.0  9.081660e+17        The Realist    therealist822   
4             0.0  5.260074e+08  Inquirer Business      InquirerBiz   

                                    user_description  \
0  Retired Tradesman and Progressive Anti Conserv...   
1  Life coach & trainer, Motivational speaker, Ma...   
2  Writes for @theage and @smh on technology and ...   
3  Calls it as I see it. Anti PC, SJW and VS. If ...   
4  The official Twitter account of the Inquirer G...   

                user_location      user_created_at  
0        Brisbane, Queensland  2009-11-25 09:19:45  
1  Wollongong, NSW, AUSTRALIA  2012-08-23 10:20:40  
2                   Melbourne  2006-12-11 07:38:06  
3                         NaN  2017-09-14 03:10:30  
4                 Philippines  2012-03-16 03:51:59

Se procede a descargar la librería nltk, muy util al momento de gestionar textos y realizar preprocesamiento de datos que vienen en formato texto.

In [11]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\anmam\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

Se guardan los textos descargados en el archivo en la variable textos

In [12]:
textos=datos[['full_text']]
textos.head()

full_text
0  After the climate election: shellshocked green...
1  @narendramodi @smritiirani Coverage of indian ...
2  @workmanalice Do you know if Facebook is relea...
3  @vanbadham We all understand we have a compuls...
4  Shares were mixed in Asia, with India and Aust...

Lo primero que se realiza es la tokenización de la variable texto ignorando los signos de puntuación y caracteres especiales tales como: ., ,, *, @, ;, etc; Para lo mismo se usa la función RegexpTokenizer.tokenize de la libreria NLKT, que permite ayudarnos a ignorar los signos de puntiuación y dividir los textos en cuestión por palabras. El resultado de este prcedimiento es test_tokens correspondiente a un vector de listas, en donde cada lista corresponde a las palabras que componen el tweet.

Se realiza una impresión de los primeros datos de text_tokens para comprobar que el algorito está funcionanco correctamente.

In [13]:
from nltk.tokenize import RegexpTokenizer
tokenizer = RegexpTokenizer(r'\w+')
text_tokens=textos.apply(lambda fila:tokenizer.tokenize(fila['full_text']),axis=1)

In [14]:
text_tokens.head()

0    [After, the, climate, election, shellshocked, ...
1    [narendramodi, smritiirani, Coverage, of, indi...
2    [workmanalice, Do, you, know, if, Facebook, is...
3    [vanbadham, We, all, understand, we, have, a, ...
4    [Shares, were, mixed, in, Asia, with, India, a...
dtype: object

Se descarga de la librería nlkt las "stopwords" correspondientes a palabaras muy comunes y las cuales no aportan información significativa de la variable.

In [15]:
from nltk.corpus import stopwords
nltk.download('stopwords')



[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\anmam\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

En la variable palabras se seleccionan las "stopwords" del idioma ingles, que corresponde al idioma en que está escrito el texto.

In [16]:
palabras = set(stopwords.words('english'))

Se crea una variable llamada text_tokens2 en donde se colocan en minúscula todas las palabras de las listas guardadas anteriormente en text_tokens, esto con el fin de sean posibles remover todas las stopwords de las listas.
Una vez todas las listas presentan palabras en minúsculas, se crea la variable tokens_despues la cual elimina todas las "stopwords" del idioma inglés de las listas. Se realiza la impresión de tokens_despues para comprobar que los algrítmos hayan funcionado correctamente.

In [17]:
text_tokens2=text_tokens.apply(lambda func2: [i.lower() for i in func2])
tokens_despues = text_tokens2.apply(lambda x: [i for i in x if not i in palabras])

In [18]:
tokens_despues.head()

0    [climate, election, shellshocked, green, group...
1    [narendramodi, smritiirani, coverage, indian, ...
2    [workmanalice, know, facebook, releasing, elec...
3    [vanbadham, understand, compulsory, preference...
4    [shares, mixed, asia, india, australia, leadin...
dtype: object

Posteriormente se realiza una técnica que en el lenguaje común se conoce como stemming, la técnica consiste en reemplazar cada palabra por su raíz, por ejemplo, las palabras fishing, fisher, y fish quedarían todas bajo la palabra fish.

Las listas después de realizar stemming se guardan en la variable result1.

In [19]:
from nltk.stem.porter import PorterStemmer
p1 = PorterStemmer()
result1=tokens_despues.apply(lambda x: [p1.stem(i) for i in x])


In [20]:
result1.head()

0    [climat, elect, shellshock, green, group, rema...
1    [narendramodi, smritiirani, coverag, indian, e...
2    [workmanalic, know, facebook, releas, elect, p...
3    [vanbadham, understand, compulsori, prefer, sy...
4    [share, mix, asia, india, australia, lead, gai...
dtype: object

Una vez realizado el stemming a las variables resultantes se aplica la lemmatization, la lemmatization corresponde a reemplazar cada palabra con la palabra de origen, por ejemplo la palabra better se reemplazaría por good. Para realizar la lemmatization se utiliza la función WordnetLemmatizer.lemmatize de la librería NLTK.

Los resultados se encuentran en la variable result2

In [21]:
nltk.download('wordnet')
from nltk.stem import WordNetLemmatizer 
lemmatizer = WordNetLemmatizer() 
result2=result1.apply(lambda x: [lemmatizer.lemmatize(i) for i in x])

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\anmam\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [22]:
result2.head()

0    [climat, elect, shellshock, green, group, rema...
1    [narendramodi, smritiirani, coverag, indian, e...
2    [workmanalic, know, facebook, releas, elect, p...
3    [vanbadham, understand, compulsori, prefer, sy...
4    [share, mix, asia, india, australia, lead, gai...
dtype: object

Una vez preprocesada la información del texto se crea procede a procesar la información de las etiquetas. Para este ejercicio en particular la variable de salidas será 0 si no hubo retweets y 1 si se presentaron retweets. Para lo mismo se utiliza la función numpy.gt, en la cual se marca como 0 cuando no hubo ningún retweet y 1 cuando se presentaron retweets. Los resultados se guardan en la variable respuestas.

In [23]:
respuestas= datos[['retweet_count']].gt(0).astype(np.short)

Una vez procesadas las respuestas se procede a verificar que otras variables serían útiles para el caso de estudio.
Para lo mismo se consideran las variables que se tienen nombre de usuario, fecha de creación, fecha de creación usuario, localización, nombre en pantalla.

Se considera intuitivamente que no hay razón  para pensar que por el nombre de usuario o localización se realicen más retweet, si bien es cierto que ciertos nombres de personas famosas tienen mayor probabilidad de ser retweeteadas, dado el alcance del ejercicio y que los tweets únicamente tienen en cuenta 10 días de historia, no se tendrá en cuenta este parámetro en las variables explicativas.

La localización tampoco se tendrá en cuenta como variable, ya que no existe una razón intuitiva para pensar que en ciertos lugares que se realice el tweet haya mayor probabilidad de ser retweeteado, el lugar de origen del tweet no debería ser un factor determinante y solo agregaría ruido al modelo.

La hora y el día en que se realice el tweet si se considera un factor determinante, ya que las personas estan más expuestas a twiter a ciertas horas del días y ciertos días a la semana, por lo tanto la probabilidad de retweet, puede estar determinada por la visibilidad del tweet. 

Dado que únicamente se cuentan con 10 días de historia únicamente se considerará la hora del día como un factor determinante y no el día de la semana, por simplicidad se ignorarán los segundos en el cálculo de la hora.
La hora del día se calcularpa como Hora militar*60 + minutos. Para lo mismo se creará una función la sual dado un String extraiga la hora en el formato especificado.

Es de resaltar que debido a que hay datos que carecen de hora en la base, se incluirá en la función que cuando no sea posible extaer la hora se generé un np.nan.


In [24]:
from datetime import datetime
def funcionfecha(y):
    try:
        x=y.to_string()
        datetime_object = datetime.strptime(x, 'created_at    %Y-%m-%d %H:%M:%S')
        return(datetime_object.hour*60+datetime_object.minute)
    except:
        return np.nan



Se aplica la función creada (funcionfecha) al campo "created_at" de los datos, para extraer las horas de cada una de las observaciones, los resultados se almacenan en la variable horasa.

In [25]:
horasa=datos[['created_at']].apply(funcionfecha,axis=1)


Se procede a dividir los datos obtenidos en entrenamiento y prueba, dado el volumen suficiente de la información y las recomendaciones de la literatura, se seleccionara 20% como datos de prueba y 80% como datos de entrenamiento.

Es así que se crea un dataframe con los datos procesados de texto, hora y existencia de retweets, el arreglo resultante se llamará datosfinal.

Usando la funcion model_selection de sklearn se realiza la partición de los datos en X_train, X_test, y_train, y_test

In [26]:
datosfinal=pd.concat([result2,horasa,respuestas],axis=1)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(datosfinal[[0,1]], datosfinal.retweet_count, test_size=0.2)
print (X_train.shape, y_train.shape)
print (X_test.shape, y_test.shape)

X_train

(146703, 2) (146703,)
(36676, 2) (36676,)


0       1
35111   [bright, side, cancer, chariti, expect, influx...   118.0
179673  [abctv, abccompass, watch, kumi, taguchi, late...   568.0
103060                     [auspol, http, co, btpyxpfx7n]  1376.0
62430   [look, like, hung, parliament, mojonewsau, aus...   668.0
176772  [worst, part, elect, tri, figur, put, last, ua...   799.0
39349          [smh, jacquelinemaley, politician, auspol]   679.0
4598    [australian, famili, guy, emerg, new, conserv,...   430.0
1752    [australia, elect, outcom, incred, shock, lect...    11.0
106443  [braddonvot, bassvot, lyonvot, abchobart, hoba...   574.0
51759   [dear, white, cishet, babi, boomer, racist, ra...   897.0
129619  [oh, good, graciou, cinema, neg, ad, bill, lnp...    13.0
172957   [guardianau, anti, lnp, propaganda, guy, ausvot]   109.0
68039   [mp, warren, entsch, concern, young, girl, see...   570.0
177246  [jim, molan, rogu, senat, campaign, label, sab...   338.0
167908  [happpyelectionday, ausvot, democracysausag, h...   217.0
87844   [vale, bob, australian, love, rememb, solidar,...   119.0
173260  [4, 05m, vote, cast, earli, vote, centr, cob, ...  1356.0
68587   [anyway, sorri, guy, elect, night, im, gonna, ...   549.0
144604  [mani, candid, pull, earli, elect, never, pre,...  1434.0
179149  [today, torr, strait, island, lodg, complaint,...  1413.0
39990   [lilydambrosiomp, seem, rural, seat, well, one...   596.0
157006  [today, believ, au, opinion, poll, say, differ...   709.0
97288   [3, elect, 3, year, amp, imagin, 1975, murdoch...   421.0
42315   [sum, auspol, ausvot, especi, clivepalm, amp, ...   343.0
573     [trade, australia, morrison, secur, major, sha...   394.0
156246  [antonygreen, point, poll, methodolog, chang, ...   727.0
61041   [whelp, look, like, anoth, 3, year, exil, aust...   689.0
64341   [easi, tonyabbottmhr, rather, loser, quitter, ...   637.0
66878   [want, labor, minor, gov, ff, ride, home, folk...   593.0
169071                              [happi, vote, ausvot]   103.0
...                                                   ...     ...
57932        [gobsmack, australia, absolut, cook, auspol]   734.0
30532   [joeltooley, describ, report, lindseybev, via,...   633.0
68529    [oz, fckd, ausvot, auspol, http, co, p0krqmxuwd]   552.0
153637  [billshortenmp, conced, elect, recontest, lead...   819.0
61016   [mayb, consol, one, bad, one, lose, let, see, ...   689.0
154108  [propos, new, unit, state, australia, ausvot, ...   802.0
64692   [tell, ya, letter, mum, got, target, elderli, ...   633.0
166713  [aapmick, peter, dutton, say, organis, send, l...   366.0
16235   [feder, elect, join, 7new, anchor, michaelush,...   330.0
53550   [interest, see, malcolm, turnbul, respons, rea...   832.0
98838   [bernardkean, swannyqld, turn, abc24, cheer, v...   281.0
24676   [vote, paulin, hanson, one, nation, want, vote...   242.0
95467   [aussie_oliv, inde, abc730, w, scottmorrisonmp...   579.0
154123  [bill, shorten, arriv, australianlabor, headqu...   801.0
40749   [quick, look, poll, place, level, 2pp, result,...   512.0
20680   [ask, australian, elect, said, australia, need...   117.0
111892  [liberalau, scottmorrisonmp, joshfrydenberg, a...  1436.0
65372   [one, thing, certain, way, track, poll, eg, ne...   623.0
80363   [join, birdlifeoz, call, elect, need, stronger...  1258.0
99512   [funni, clive, palmer, spent, 20, million, get...   235.0
109705  [baxterpeterba, right, high, commiss, london, ...   223.0
99719   [zalisteggal, acoss, worri, zalisteggal, fix, ...   220.0
75830   [front, gate, moone, pond, west, p, billshorte...   103.0
15119   [top, stori, de, brandmarkespana, australian, ...   560.0
46147   [phillipadams_1, lie, corrupt, paralysi, disun...    55.0
81474   [auspol, obvious, pas, bob, hawk, put, lnp, qu...   782.0
102307  [auspol, australiavotes2019, http, co, 2cm8me3...     7.0
75410   [vote, hand, volunt, mani, candid, one, messag...   126.0
30253   [c, mon, debfrecklington, lnp, colleagu, conde...   670.0
23168   [mecardoanalysi, elect

Dado a que se tienen datos faltantes en el campo de hora, se crean variables nuevas de tal manera que los datos faltantes tanto en los datos de observación como en los datos de prueba correspondan a la media de los demás datos, esto con el fin de que los algorítmos sean capaces de procesar la información, y de acuerdo a las recomendaciones de varios autores del tema, que sugieren reemplazar por la media o la mediana.

Las variables obtenidas de hora después de hacer este reemplazo corresponden a horastrain y horastest.

In [27]:
import numpy.ma as ma
horastrain=np.where(np.isnan(X_train[[1]]), ma.array(X_train[[1]], mask=np.isnan(X_train[[1]])).mean(axis=0), X_train[[1]]) 
horastest=np.where(np.isnan(X_test[[1]]), ma.array(X_test[[1]], mask=np.isnan(X_test[[1]])).mean(axis=0), X_test[[1]]) 
horastrain = horastrain.reshape(len(horastrain), 1)
horastest = horastest.reshape(len(horastest), 1)

Para procesar la información de las listas de palabras de los textos de los tweets, se utilizará el algoritmo TF-IDF, el cual asigna a cada palabra un valor dependiendo del número de veces que aparece en el texto y la frecuencia de aparición en los textos. Es así que el valor TF-IDF por palabra será calculado como (Número de veces que la palabra aparece en el tweet)/(Número de palabras en el tweet)*ln(Número de documentos/número de tweets que contiene la palabra)

Para realizar este procedimiento y cuantificar los textos de los tweets se utilizará la función TfidfVectorizer de sklearn, esta permite hallar el TF_IDF de todas las palabras de todos los tweets  tanto en los datos de entrenamiento como en los datos de prueba. Esto permite extraer el nivel de rareza de las palabras que presentacada tweet.

En primer lugar se calcula el valor TF-IDF para los datos de entranamiento y se guarda en la matriz datos train, posteriormente se extrae el vocabulario usado en los datos de entrenamiento, con el cual se calculó el TF-IDF, para usar el mismo diccionario en los datos de prueba. Se considera que dado que los datos de entrenamiento son suficientemente amplios, la mayoria de las palabras de los datos de prueba estará en este diccionario. Este diccionario se guardará en la variable aa.

Posteriormente se calcula el TF-IDF para los datos de prueba basados en las palabras de los datos de entrenamiento. El resultado se guardará en la variable datostest.

In [28]:
from sklearn.feature_extraction.text import TfidfVectorizer
def funciondummy(doc):
    return doc
tfidf = TfidfVectorizer(
    analyzer='word',
    tokenizer=funciondummy,
    preprocessor=funciondummy,
    token_pattern=None)
Xftr=X_train[[0]].iloc[:,0]
Xftt=X_test[[0]].iloc[:,0]

In [109]:
respuesta=tfidf.fit(Xftr)
datostrain=respuesta.transform(Xftr)


In [113]:
aa=tfidf.vocabulary_.keys()
print(len(aa))
datostrain.shape


147099


(146703, 147099)

In [114]:
tfidf = TfidfVectorizer(
    analyzer='word',
    tokenizer=funciondummy,
    preprocessor=funciondummy,
    token_pattern=None, vocabulary=aa)

In [115]:
respuesta=tfidf.fit(Xftt)
datostest=respuesta.transform(Xftt)

In [116]:
print(datostrain.shape)
datostest.shape

(146703, 147099)


(36676, 147099)

In [117]:
print(tfidf.get_feature_names)

<bound method CountVectorizer.get_feature_names of TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.float64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), norm='l2',
        preprocessor=<function funciondummy at 0x000002B7AE812950>,
        smooth_idf=True, stop_words=None, strip_accents=None,
        sublinear_tf=False, token_pattern=None,
        tokenizer=<function funciondummy at 0x000002B7AE812950>,
        use_idf=True,
        vocabulary=dict_keys(['bright', 'side', 'cancer', 'chariti', 'expect', 'influx', 'donat', 'frankingcredit', 'refund', 'need', 'coupl', 'billion', 'match', 'labor', 'plan', 'auspol', 'ausvot', 'abctv', 'abccompass', 'watch', 'kumi', 'taguchi', 'latest', 'present', 'brilliant', 'obviou', 'virtu', 'par...guym', 'anqe6m467d', 'quandari', '2cm8me3hpm', 'xo65gj2j5u', 'nibpdzseka', 'mecardoanalysi', '255']))>


Se unen los valores obtenidos por TF-IDF con el parametro de la hora tanto para los datos de entrenamiento como para los de prueba. Dado que los resultados de TF-IDF son una sparse matrix, se utiliza la función hstack. 
La matriz resultante para los datos de entrenamiento es Xfinaltrain y para los de prueba Xfinaltest

In [118]:
import scipy.sparse as spar
Xfinaltrain=spar.hstack((datostrain,horastrain))
Xfinaltest=spar.hstack((datostest,horastest))

Con el fin de reducir la dimensionalidad de los datos de entrada y poder procesar la información con los resursos y tiempo disponibles, se usará el metodo TruncatedSVD, se intentó usar el método PCA, y estandarizar los datos, pero dado a que se están usando sparse matrix, esto no fue posible con un computador con los que se contaba.

El metodo TruncatedSVD tiene la ventaja de que las entradas al mismo no deben estar estandarizadas para su uso y es utilizado frecuentemente con sparse matrix, a pesar de que la literatura de análisis de texto sugiere usar 100 dimensiones como salida, para este caso en particular se utilizarán únicamente 10, de manera que la información se pueda procesar más rapidamente cuando se esten implementando los algoritmos de entrenamiento. Los resultados se guardan en las matrices X_projtrain para los datos de entrenamiento y X_projtestpara los datos de prueba.

In [119]:
from sklearn.decomposition import TruncatedSVD
model = TruncatedSVD(n_components=10).fit(Xfinaltrain)
X_projtrain = model.transform(Xfinaltrain)
model = TruncatedSVD(n_components=10).fit(Xfinaltest)
X_projtest = model.transform(Xfinaltest)

De esta manera los resultados para los datos de entrenamiento X_projtrain y para los datos de prueba X_projtest, ya se encuentran listos para usar en un algoritmo de ML.

In [120]:
y_test

51735     1
123772    0
165888    0
15266     0
134395    0
96736     0
115754    1
178744    0
127185    0
144997    0
120456    0
158540    1
120850    0
118163    1
82414     1
28075     0
2919      0
123801    0
166072    1
75884     0
3984      0
163217    0
36206     0
48424     0
28022     0
93829     0
13600     0
75741     0
36882     0
94308     1
         ..
65464     1
90881     0
75786     0
85570     0
81808     0
89624     1
127492    1
32051     1
73732     0
13125     0
45875     0
2739      0
120546    0
73026     1
153986    0
117757    1
16508     0
141911    0
37367     0
22159     0
123503    0
131306    1
176720    0
57990     0
153175    0
66810     1
13705     0
140182    0
71600     1
157116    0
Name: retweet_count, Length: 36676, dtype: int16

A continuación se realizarán diferentes algoritmos de Machine Learning que permitan determinar la probabilidad de que un tweet sea retweeteado con base en su contenido y la hora en que fue publicado. Para realizar este procedimiento se tomaron en cuenta 3 algoritmos: Support Vector Machines, Redes Neuronales con capas ocultas y Adaboost con árboles de decisión como clasificador debil.

El alcance del proyecto únicamente se limitó a estos 3 metodos, dada los buenos resultados que los mismos han obtenido en la literatura, y por falta de tiempo y recursos tecnológicos para realizar otros metodos, no obstante se tuvieron en consideracón: Random Forest, otros algorítmos de boosting como el de gradiente estocástico y modelos paramétricos como son el Logit y el Probit.

En todos los modelos mostrados en este trabajo se realizó validación cruzada de k iteraciones que consiste en dividir los datos de entrenamiento en k muestras, tomar k-1 para entrenar el modelo y 1 para validarlo, y rotar los sets de elementos usados en cada tarea. Dadas las disposiciones tecnológicas y el tiempo, únicamente se tuvieron en cuenta 5 divisiones para la validación cruzada k=5. Para la validación cruzada en todos los metodos se utilizó la función cross_val_score.

El primer modelo que se tomó como referencia fue Support Vector Machines, aunque inicialmente la idea era probar el algoritmo con diferentes tipos de kernel (lineal, polínomico, gaussiano), sin embargo por falta de tiempo únicamente se utilizó el kernel "Radial Basis Function", el cual tiene un hiperparametro adicional gamma.

A continuación se presenta el modelo de entrenamiento de SVM, para calcular los hiperparametros C (el factor de regularización) y gamma, se realizo una pequeña Grid Search en una escala logaritmica para ambos parametros, para gamma se tomaron 6 valores entre 10e-2 y 10e3, y para C entre 10e-2 y 10e8. Se escogieron estos rangos dado a que se prefiere la simplicidad en los modelos y el algoritmo tarda bastante en su entrenamiento con la tecnología actual. Se limitó el número máximo de iteraciones a 50 para poder realizar el entrenamiento y el grid search rápidamente.

Los resultados del nivel de clasificación por cada hiperparametro se guarda en la variabe lista.

In [121]:
gamma_range = np.logspace(-2, 3, 6)
C_range = np.logspace(-2, 8, 6)


In [122]:
from sklearn.model_selection import cross_val_score
from sklearn import svm
lista=[]
for c in C_range:
    for gamma in gamma_range:
        clf = svm.SVC(kernel='rbf', C=c, gamma=gamma, max_iter=500).fit(X_projtrain, y_train)
        scores = cross_val_score(clf, X_projtest, y_test, cv=5)
        lista.append((np.mean(scores),c,gamma))
        

C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\s

C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\s

C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\s

C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\s

C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\s

C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\s

C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\sklearn\svm\base.py:244: ConvergenceWarning: Solver terminated early (max_iter=500).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)
C:\Users\anmam\Anaconda3\lib\site-packages\s

In [123]:
lista

[(0.6514888258097603, 0.01, 0.01),
 (0.6572690031753357, 0.01, 0.1),
 (0.659286722727567, 0.01, 1.0),
 (0.6583869123430806, 0.01, 10.0),
 (0.6566418915647522, 0.01, 100.0),
 (0.5948644490464269, 0.01, 1000.0),
 (0.6507254893320042, 1.0, 0.01),
 (0.6570781263115488, 1.0, 0.1),
 (0.6586868255856013, 1.0, 1.0),
 (0.6584959524406535, 1.0, 10.0),
 (0.6552514174577178, 1.0, 100.0),
 (0.5949462486475647, 1.0, 1000.0),
 (0.641618880723967, 100.0, 0.01),
 (0.6492802368334469, 100.0, 0.1),
 (0.6548151826916631, 100.0, 1.0),
 (0.6565874328382904, 100.0, 10.0),
 (0.6553604761413763, 100.0, 100.0),
 (0.5950654704646655, 100.0, 1000.0),
 (0.6391918397862006, 10000.0, 0.01),
 (0.6469624739229198, 10000.0, 0.1),
 (0.6563420340733928, 10000.0, 1.0),
 (0.65552410505686, 10000.0, 10.0),
 (0.6551696178646885, 10000.0, 100.0),
 (0.595147266338856, 10000.0, 1000.0),
 (0.6464987124355248, 1000000.0, 0.01),
 (0.6513250556057678, 1000000.0, 0.1),
 (0.6541879707231105, 1000000.0, 1.0),
 (0.40894230806762605, 10

Se observa que el mejor resultado se obtiene cuando tanto gamma como c tiene un valor de 0.01 y gamma un valor de 1, con estos resultados podría realizarse nuevamente grid_search para ajustar de mejor manera el valor de los hiperparametros, pero dado al tiempo y teconología disponibles, esto se saldrá del alcance de este ejercicio. Cabe resaltar que estos resultados pueden estar equivocos dada la limitación del máximo número de iteraciones. El nivel máximo de exito alcanzado fue 65.92%

El segundo modelo que se utilizó fue el modelo de redes neuronales con capas ocultas, para este caso únicamente se consideró hasta dos capas ocultas, ya que entre más complejo sea el modelo, más dificil la estimación y menos deseable para el análisis, además que se contaba con poco tiempo. 
El número de neuronas escogida como hiperparametro para cada una de las capas se encontró por medio de una pequeña grid search, en donde se tomaron valores entre 1 y 21  neuronas para la primera capa y entre 0 y 20 para la segunda, se limitó nuevamente el número máximo de iteraciones hasta 150. El numero de particiones para la validación cruzada continúa siendo 5.

Los resultados de elnivel de clasificación en los datos de prueba con cada par de hiperparametros se guardan en la variable listarn.

In [124]:
from sklearn.neural_network import MLPClassifier

capa1 = np.linspace(1, 21, 5)
capa2 = np.linspace(0, 20, 5)
listarn=[]
for c1 in capa1:
    for c2 in capa2:
        c1=int(c1)
        c2=int(c2)
        if c2==0:
            mlp = MLPClassifier(hidden_layer_sizes=(c1,), max_iter=150, alpha=1e-4,
                    solver='sgd', verbose=10, tol=1e-3, random_state=1,
                    learning_rate_init=.1)
        else:
            mlp = MLPClassifier(hidden_layer_sizes=(c1,c2), max_iter=150, alpha=1e-4,
                    solver='sgd', verbose=10, tol=1e-3, random_state=1,
                    learning_rate_init=.1)
        mlp.fit(X_projtrain, y_train)
        scores = cross_val_score(mlp, X_projtest, y_test, cv=5)
        listarn.append((np.mean(scores),c1,c2))

Iteration 1, loss = 0.64107652
Iteration 2, loss = 0.64094130
Iteration 3, loss = 0.64097342
Iteration 4, loss = 0.64094889
Iteration 5, loss = 0.64084724
Iteration 6, loss = 0.64086510
Iteration 7, loss = 0.64608677
Iteration 8, loss = 0.64813519
Iteration 9, loss = 0.64799941
Iteration 10, loss = 0.64808019
Iteration 11, loss = 0.64813369
Iteration 12, loss = 0.64809197
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.64189611
Iteration 2, loss = 0.64132938
Iteration 3, loss = 0.64144143
Iteration 4, loss = 0.64138102
Iteration 5, loss = 0.64139897
Iteration 6, loss = 0.64147340
Iteration 7, loss = 0.64132093
Iteration 8, loss = 0.64127771
Iteration 9, loss = 0.64139732
Iteration 10, loss = 0.64124245
Iteration 11, loss = 0.64162450
Iteration 12, loss = 0.64146569
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.64187148
Iteration 2, loss = 0.64141692
Iteratio

Iteration 8, loss = 0.64141625
Iteration 9, loss = 0.64148450
Iteration 10, loss = 0.64151899
Iteration 11, loss = 0.64146543
Iteration 12, loss = 0.64152428
Iteration 13, loss = 0.64154152
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.64140263
Iteration 2, loss = 0.64114274
Iteration 3, loss = 0.64112250
Iteration 4, loss = 0.64105361
Iteration 5, loss = 0.64095236
Iteration 6, loss = 0.64097194
Iteration 7, loss = 0.64097827
Iteration 8, loss = 0.64087638
Iteration 9, loss = 0.64087841
Iteration 10, loss = 0.64100570
Iteration 11, loss = 0.64091057
Iteration 12, loss = 0.64089233
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.64176407
Iteration 2, loss = 0.64168295
Iteration 3, loss = 0.64160554
Iteration 4, loss = 0.64170127
Iteration 5, loss = 0.64164354
Iteration 6, loss = 0.64181399
Iteration 7, loss = 0.64167994
Iteration 8, loss = 0.64161850
Iterati

Iteration 1, loss = 0.65890923
Iteration 2, loss = 0.64418172
Iteration 3, loss = 0.64431758
Iteration 4, loss = 0.64404366
Iteration 5, loss = 0.64416131
Iteration 6, loss = 0.64417918
Iteration 7, loss = 0.64406349
Iteration 8, loss = 0.64406346
Iteration 9, loss = 0.64407896
Iteration 10, loss = 0.64414947
Iteration 11, loss = 0.64404862
Iteration 12, loss = 0.64424023
Iteration 13, loss = 0.64424836
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.64815892
Iteration 2, loss = 0.64149880
Iteration 3, loss = 0.64131918
Iteration 4, loss = 0.64142296
Iteration 5, loss = 0.64143095
Iteration 6, loss = 0.64135967
Iteration 7, loss = 0.64136620
Iteration 8, loss = 0.64135104
Iteration 9, loss = 0.64128964
Iteration 10, loss = 0.64131326
Iteration 11, loss = 0.64137592
Iteration 12, loss = 0.64130142
Iteration 13, loss = 0.64141057
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iterat

Iteration 11, loss = 0.64320541
Iteration 12, loss = 0.64291079
Iteration 13, loss = 0.64319550
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.67866811
Iteration 2, loss = 0.64302992
Iteration 3, loss = 0.64308357
Iteration 4, loss = 0.64288229
Iteration 5, loss = 0.64298504
Iteration 6, loss = 0.64302436
Iteration 7, loss = 0.64291326
Iteration 8, loss = 0.64296152
Iteration 9, loss = 0.64293518
Iteration 10, loss = 0.64282603
Iteration 11, loss = 0.64310078
Iteration 12, loss = 0.64287396
Iteration 13, loss = 0.64312412
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.69218901
Iteration 2, loss = 0.64437090
Iteration 3, loss = 0.64452347
Iteration 4, loss = 0.64451568
Iteration 5, loss = 0.64421745
Iteration 6, loss = 0.64460253
Iteration 7, loss = 0.64438554
Iteration 8, loss = 0.64440405
Iteration 9, loss = 0.64446935
Iteration 10, loss = 0.64439828
Iterat

Iteration 5, loss = 2.58101651
Iteration 6, loss = 2.58067556
Iteration 7, loss = 2.58034068
Iteration 8, loss = 2.58029290
Iteration 9, loss = 2.58006526
Iteration 10, loss = 2.57963138
Iteration 11, loss = 2.57929763
Iteration 12, loss = 2.57909031
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 2.57001588
Iteration 2, loss = 2.62077745
Iteration 3, loss = 2.62059149
Iteration 4, loss = 2.62024251
Iteration 5, loss = 2.61991902
Iteration 6, loss = 2.61952953
Iteration 7, loss = 2.61920286
Iteration 8, loss = 2.61902641
Iteration 9, loss = 2.61870669
Iteration 10, loss = 2.61832593
Iteration 11, loss = 2.61829332
Iteration 12, loss = 2.61778072
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 2.17621482
Iteration 2, loss = 2.18789345
Iteration 3, loss = 2.18772223
Iteration 4, loss = 2.18736516
Iteration 5, loss = 2.18720788
Iteration 6, loss = 2.18702269
Iteratio

Iteration 2, loss = 0.64573935
Iteration 3, loss = 0.64589351
Iteration 4, loss = 0.64576538
Iteration 5, loss = 0.64580237
Iteration 6, loss = 0.64565078
Iteration 7, loss = 0.64603230
Iteration 8, loss = 0.64522163
Iteration 9, loss = 0.64571935
Iteration 10, loss = 0.64561643
Iteration 11, loss = 0.64564052
Iteration 12, loss = 0.64555169
Iteration 13, loss = 0.64556172
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.70794272
Iteration 2, loss = 0.64619114
Iteration 3, loss = 0.64589483
Iteration 4, loss = 0.64595652
Iteration 5, loss = 0.64591146
Iteration 6, loss = 0.64567543
Iteration 7, loss = 0.64597947
Iteration 8, loss = 0.64568677
Iteration 9, loss = 0.64578218
Iteration 10, loss = 0.64549781
Iteration 11, loss = 0.64569571
Iteration 12, loss = 0.64558953
Iteration 13, loss = 0.64557341
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 0.69602664
Iterat

Iteration 1, loss = 2611.86736614
Iteration 2, loss = 2663.40233652
Iteration 3, loss = 2661.44600631
Iteration 4, loss = 2659.49120648
Iteration 5, loss = 2657.53778924
Iteration 6, loss = 2655.58586980
Iteration 7, loss = 2653.63531724
Iteration 8, loss = 2651.68621885
Iteration 9, loss = 2649.73854666
Iteration 10, loss = 2647.79240516
Iteration 11, loss = 2645.84751009
Iteration 12, loss = 2643.90413577
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 2661.14997262
Iteration 2, loss = 2957.05147834
Iteration 3, loss = 2956.61571531
Iteration 4, loss = 2956.17987481
Iteration 5, loss = 2955.74432885
Iteration 6, loss = 2955.30841649
Iteration 7, loss = 2954.87326100
Iteration 8, loss = 2954.43754339
Iteration 9, loss = 2954.00217085
Iteration 10, loss = 2953.56658799
Iteration 11, loss = 2953.13136529
Iteration 12, loss = 2952.69609159
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping

Iteration 8, loss = 2211.22018624
Iteration 9, loss = 2210.89434654
Iteration 10, loss = 2210.56845092
Iteration 11, loss = 2210.24260346
Iteration 12, loss = 2209.91693445
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 1700.88736699
Iteration 2, loss = 1889.99398733
Iteration 3, loss = 1889.71546296
Iteration 4, loss = 1889.43713495
Iteration 5, loss = 1889.15871413
Iteration 6, loss = 1888.88030289
Iteration 7, loss = 1888.60196665
Iteration 8, loss = 1888.32383706
Iteration 9, loss = 1888.04551683
Iteration 10, loss = 1887.76726759
Iteration 11, loss = 1887.48925796
Iteration 12, loss = 1887.21099877
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 18.43476217
Iteration 2, loss = 18.75829789
Iteration 3, loss = 18.74500528
Iteration 4, loss = 18.73485766
Iteration 5, loss = 18.72438019
Iteration 6, loss = 18.71119631
Iteration 7, loss = 18.69791269
Iteration 8,

Iteration 2, loss = 498.52208578
Iteration 3, loss = 498.44851072
Iteration 4, loss = 498.37517736
Iteration 5, loss = 498.30170330
Iteration 6, loss = 498.22854194
Iteration 7, loss = 498.15508677
Iteration 8, loss = 498.08168341
Iteration 9, loss = 498.00847747
Iteration 10, loss = 497.93512518
Iteration 11, loss = 497.86187787
Iteration 12, loss = 497.78848428
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 522.08570898
Iteration 2, loss = 579.79896481
Iteration 3, loss = 579.71364092
Iteration 4, loss = 579.62822080
Iteration 5, loss = 579.54273048
Iteration 6, loss = 579.45744719
Iteration 7, loss = 579.37213264
Iteration 8, loss = 579.28682936
Iteration 9, loss = 579.20150027
Iteration 10, loss = 579.11642204
Iteration 11, loss = 579.03105505
Iteration 12, loss = 578.94564411
Training loss did not improve more than tol=0.001000 for 10 consecutive epochs. Stopping.
Iteration 1, loss = 431.67833850
Iteration 2, loss = 47

In [125]:
listarn

[(0.6598047793723172, 1, 0),
 (0.6598047793723172, 1, 5),
 (0.6598047793723172, 1, 10),
 (0.6598047793723172, 1, 15),
 (0.6598047793723172, 1, 20),
 (0.6598047793723172, 6, 0),
 (0.6598047793723172, 6, 5),
 (0.6598047793723172, 6, 10),
 (0.6598047793723172, 6, 15),
 (0.6598047793723172, 6, 20),
 (0.6598047793723172, 11, 0),
 (0.6598047793723172, 11, 5),
 (0.6598047793723172, 11, 10),
 (0.6598047793723172, 11, 15),
 (0.6598047793723172, 11, 20),
 (0.6598047793723172, 16, 0),
 (0.6598047793723172, 16, 5),
 (0.6598047793723172, 16, 10),
 (0.659859304999362, 16, 15),
 (0.6598047793723172, 16, 20),
 (0.6598047793723172, 21, 0),
 (0.6598047793723172, 21, 5),
 (0.6598047793723172, 21, 10),
 (0.6598047793723172, 21, 15),
 (0.6597502388759986, 21, 20)]

Se observa que el valor que minimiza el error corresponde a una red neuronal con 16 neuronas en la primera capa y 15 en la segunda. No obstante dado que el valor no cambia mucho al aumentar el número de capas y estas aumentan complejidad al modelo, se considerará como la mejor alternativa de este modelo el correspondiente 1 capa oculta con 1 red neuronal, con un nivel de éxito de 65.98%.

Finalmente el último modelo utilizado corresponde a un algoritmo de boosting (Adaboost), basado en un clasificador debil (Árboles de decisión),el numero de modelos iterativos del clasificador debil entrenado se utilizó como parametro en un gridsearch entre 15 y 50, con el fin de no utilizar algoritmos que tardaran muco en su ejecución. El valor de particiones k para la validación cruzada se mantuvo en 5 para este caso.

In [126]:
from sklearn.ensemble import AdaBoostClassifier
itera=np.linspace(15,50,8)
listab=[]
for i in itera:
    i=int(i)
    abc = AdaBoostClassifier(n_estimators=i,learning_rate=1)
    model = abc.fit(X_projtrain, y_train)
    scoresBoost = cross_val_score(model, X_projtest, y_test, cv=5)
    print(scoresBoost)
    listab.append((np.mean(scoresBoost),i))

[0.6598964  0.66044166 0.66107703 0.66244035 0.65734933]
[0.657988   0.66330425 0.66039536 0.66066803 0.65925825]
[0.66125954 0.66316794 0.66339468 0.66421268 0.6588492 ]
[0.66316794 0.66207743 0.66175869 0.66325835 0.6593946 ]
[0.66248637 0.66044166 0.66230402 0.66394001 0.66143987]
[0.66166848 0.66139586 0.66257669 0.66298569 0.66116717]
[0.6608506  0.65880589 0.66380368 0.66571234 0.66184892]
[0.66289531 0.66112323 0.66271302 0.66530334 0.66157622]


In [127]:
listab

[(0.660240954635937, 15),
 (0.6603227802588109, 20),
 (0.66217680768138, 25),
 (0.6619314014788048, 30),
 (0.6621223861443198, 35),
 (0.6619587758098536, 40),
 (0.6622042860922774, 45),
 (0.6627222237950001, 50)]

El mejor algoritmo encontrado corresponde a 50 parametros del algoritmo Adaboost con arboles de decisión con un porcentaje de éxito del 66.27%. Podría encontrarse probablemente mayor éxito al aumentar los parametros, pero esto complciaria el modelo lo cual tampoco es deseable en el alcance de este proyecto.

De esta manera se encontró un algoritmo que permite predecir la probabilidad de retweet con mayor éxito (66.27%) corresponde a Adaboost con arboles de decisión como clasificador debil, y un valor de como número de parametros de 50, el cual puede ser usado con el modelo encontrado y la función predict de sklearn (model.predict))

De otro lado se encuentra que los tres algoritmos analizados presentaron un porcentaje de exito muy similar y relativamente bajo, esto se debe principalmente a que se redujo la dimensionalidad de las variables en gran medida y se limitaron las iteraciones por algoritmo.